Set missing default values for EC2 inventory

jtyr opened the pull request:
##### SUMMARY
This PR is setting missing default values matching the `ec2.ini` file. This makes the `ec2.py` script to run even without the `ec2.ini` file in default settings. It also allows the user to create his/her own `ec2.ini` file only with options which he/she wants to change against the default values.

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
`ec2.py` inventory

##### ANSIBLE VERSION
It's not part of the distribution.

jtyr committed:
Set missing default values for EC2 inventory

samdoran commented:
I like the idea of this.

When `ec2.ini` is absent, or there is an `ec2.ini` file without an `ec2` section, it throws an exception:

```
Traceback (most recent call last):
  File "./ec2.py", line 1672, in <module>
    Ec2Inventory()
  File "./ec2.py", line 181, in __init__
    self.read_settings()
  File "./ec2.py", line 495, in read_settings
    pattern_include = config.get('ec2', 'pattern_include')
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ConfigParser.py", line 607, in get
    raise NoSectionError(section)
ConfigParser.NoSectionError: No section: 'ec2'
```

You'll need to account for the absence of the file completely, or an `ec2.ini` file missing an `ec2` section.

jtyr committed:
Make it run even with no ec2.ini file

jtyr commented:
@samdoran Thanks for your comment. Should be fixed in the last commit.

jtyr committed:
Fixing INI file reading

s-hertel reviewed (commented):
Works for me.

ryansb commented on contrib/inventory/ec2.py:
```diff
@@ -253,12 +254,18 @@ def read_settings(self):
         # Regions
         self.regions = []
-        configRegions = config.get('ec2', 'regions')
+        if config.has_option('ec2', 'regions'):
```
Instead of doing this everywhere, on instantiation of the `ConfigParser` (or SafeConfigParser) you can pass a dictionary of defaults, so all the `config.get.....` calls need no alteration, you just need to build a dictionary of all the defaults (for `regions`, `regions_exclude`, and so forth. It would mean we don't have defaults sprinkled all over the code, and probably be net fewer lines.

jtyr replied:
That would be ideal solution but it would take much more effort. I have just implemented the missing defaults in the same style like the others. In the next step (and PR) we can do what you are proposing. If the code works for you, let's merge it now and discuss other changes in the IRC channel.

samdoran committed:
Refactor how defaults are handeled

Define defaults in a dictionary and use .get rather than if statements with has_option

samdoran commented:
@ryansb @s-hertel Please give this a review.

I refactored all the settings and included a `DEFAULTS` variable.

Here's my test matrix:

✅ no ini file
✅ empty ini file
✅ ini file with blank ec2 section
✅ ini file with blank credentials section
✅ ini file with blank ec2 and credentials sections
✅ ini file with a minimal set of options set in the ec2 section

s-hertel commented on contrib/inventory/ec2.py:
```diff
@@ -409,88 +437,40 @@ def read_settings(self):
-                setattr(self, option, config.getboolean('ec2', option))
-            else:
-                setattr(self, option, True)
+            setattr(self, option, config.getboolean('ec2', option))
```
Looks good to me!

This is working as expected.

A couple minor things I noticed:
* Using STS credentials doesn't work with dynamic inventory (it doesn't work on devel either).
* I'm also wondering about this bit. I'm reading the removed lines 459-462 as 'if this group_by_thing doesn't have the option provided it defaults to true' - if the file doesn't exist then it seems like the behavior is changing a little. (such as, group_by_aws_account is false by default in the ini but if it isn't provided it defaults to true here; now it defaults to false always). I think that's okay though.

samdoran replied:
I transfered the defaults in the `ini` file to the `DEFAULTS` variable. The two `group_by` options that are `False` in the `ini` file are false in `DEFAULTS`. All other `group_by` options are `True`.

See lines [174](https://github.com/ansible/ansible/pull/28375/files#diff-c2e371537339bbab5d9d0cc5d6cb7adcR174) and [180](https://github.com/ansible/ansible/pull/28375/files#diff-c2e371537339bbab5d9d0cc5d6cb7adcR180).

s-hertel replied:
Yeah, I see that you kept the behavior specified by default in the ini. I just meant that before, if the group_by option was not provided those by-default Falses would default-default to True by omission (as opposed to now, when it is always False by omission) on the deleted line 462. Just a weird corner case.

samdoran replied:
Oh, I had not connected those dots. I suppose that begs the question "What should the default be?". Hopefully sticking with what was defined in the `ini` file will be ok. I think that was effectively the default anyway since the script before would not run without an `ini` file.

Thanks for the feedback.

samdoran commented:
@ryansb Can you re-review please?

jtyr commented:
@ryansb please could you review again to unblock this to be merged?

s-hertel commented:
@jtyr Do you mind rebasing this, please?

jtyr committed:
Removing double keys and fixing logic for instance_filter

jtyr commented on contrib/inventory/ec2.py:
```diff
@@ -201,8 +200,8 @@
     'regions_exclude': 'us-gov-west-1, cn-north-1',
     'replace_dash_in_groups': 'True',
     'replace_dash_in_groups': 'True',
-    'route53': 'False',
```
This is just alphabetical re-ordering.

jtyr commented on contrib/inventory/ec2.py:
```diff
@@ -178,7 +178,6 @@
     'group_by_elasticache_replication_group': 'True',
     'group_by_instance_id': 'True',
     'group_by_instance_state': 'False',
-    'group_by_instance_state': 'True',
```
The default is `False` in the `ec2.ini` file so removing this one.

jtyr commented on contrib/inventory/ec2.py:
```diff
@@ -474,13 +473,7 @@ def read_settings(self):
             filters = config.get('ec2', 'instance_filters')
 
             if self.stack_filters and '&' in filters:
-                instance_filter = instance_filter.strip()
```
The `instance_filter` doesn't exist at this point so the code would fail here.

jtyr commented on contrib/inventory/ec2.py:
```diff
@@ -474,13 +473,7 @@ def read_settings(self):
-                instance_filter = instance_filter.strip()
-                if not instance_filter or '=' not in instance_filter:
-                    continue
                 self.fail_with_error("AND filters along with stack_filter enabled is not supported.\n")
```
The `fail_with_error()` exists so anything beyond this point can be removed.

jtyr reviewed (commented):
@samdoran Please review. See my comments for explanation.

jtyr committed:
Removing one more doubled key

samdoran commented:
@jtyr Sorry about the bad merge. That's what I get for doing a rebase at the end of the day. 😊

Thanks for getting it straight.
